### PR TITLE
fixed marker generation in charuco board

### PR DESCRIPTION
**Pull request by AleksandrPanov**
When generating  Charuco board markers in `generateImage()`, a problem occurs as in https://github.com/opencv/opencv/issues/24806, https://github.com/opencv/opencv/pull/24873:

In low resolution:
charucoImg_before_fix2
In medium resolution:
charucoImg_before_fix

This PR fixed this problems:
charucoImg_after_fix2
charucoImg_after_fix

The test checks the inner and outer borders of the Aruco markers. In the outer border of Aruco marker, all pixels should be white. In the inner border of Aruco marker, all pixels should be black.
image

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/objdetect/test/test_charucodetection.cpp [hunk 1]
```diff
@@ -817,7 +824,51 @@ TEST(CharucoBoardGenerate, issue_24806)
...
+        Mat markerImg = arucoZone(borderRect);
+        Mat markerInnerBorder = markerImg;
+        markerInnerBorder.adjustROI(-1, -1, -1, -1);
+        Mat markerOuterBorder = markerImg;
+        markerOuterBorder.adjustROI(1, 1, 1, 1);
+        const int whitePixOnMarkerBorder = countNonZero(markerImg) - countNonZero(markerInnerBorder);
+        ASSERT_EQ(whitePixOnMarkerBorder, 0);
+        const int whitePixOnOuterMarkerBorder = countNonZero(markerOuterBorder) - countNonZero(markerImg);
+        const int goldWhitePixOnOuterMarkerBorder = 2*markerOuterBorder.cols + 2*markerOuterBorder.rows - 4;
+        ASSERT_EQ(whitePixOnOuterMarkerBorder, goldWhitePixOnOuterMarkerBorder);
```
Comment (asmorkalov) [inline modules/objdetect/test/test_charucodetection.cpp:870]: I propose to extract checks as a local function or two to make the logic obvious. It's very hard to understand the code without attached image. The attached image is not visible, if you just read code in git or try to fix some bug.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/objdetect/test/test_charucodetection.cpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/objdetect/test/test_charucodetection.cpp]: I made the test simple:

```
        //The test checks the inner and outer borders of the Aruco markers.
        //In the inner border of Aruco marker, all pixels should be black.
        //In the outer border of Aruco marker, all pixels should be white.

        Mat markerImg = arucoZone(borderRect);
        bool markerBorderIsBlack = borderPixelsHaveSameColor(markerImg, 0);
        ASSERT_EQ(markerBorderIsBlack, true);

        Mat markerOuterBorder = markerImg;
        markerOuterBorder.adjustROI(1, 1, 1, 1);
        bool markerOuterBorderIsWhite = borderPixelsHaveSameColor(markerOuterBorder, 255);
        ASSERT_EQ(markerOuterBorderIsWhite, true);
```

Comment (AleksandrPanov) [reply modules/objdetect/test/test_charucodetection.cpp]: borderPixelsHaveSameColor:

```
static bool borderPixelsHaveSameColor(const Mat& image, uint8_t color) {
    for (int j = 0; j < image.cols; j++) {
        if (image.at<uint8_t>(0, j) != color || image.at<uint8_t>(image.rows-1, j) != color)
            return false;
    }
    for (int i = 0; i < image.rows; i++) {
        if (image.at<uint8_t>(i, 0) != color || image.at<uint8_t>(i, image.cols-1) != color)
            return false;
    }
    return true;
}
```

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**